### PR TITLE
refactored code

### PR DESCRIPTION
**Pull request by jhaco**

### CODE REVIEW

File: smfile_parser.py [hunk 1]
```diff
@@ -1,103 +1,95 @@
...
 
-def parse_sm(sm_file, new_file, output_dir):
-    step_dict = defaultdict(list)
-    measure         = []
-    measure_index   = 0
-
-    flag            = False
+def get_data(sm_file):
+    data = defaultdict(list)
+    flag = False
```
Comment (cpuguy96) [inline smfile_parser.py]: Rename these for better clarity

File: smfile_parser.py [hunk 2]
```diff
@@ -1,103 +1,95 @@
...
-                    line_timing = calculate_timing(measure, measure_index, step_dict['BPM'], step_dict['offset'])
-                    step_dict['notes'].extend(line_timing)
-                    measure.clear()
-                    if line.startswith(';'):
-                        output_file(new_file, step_dict, output_dir)
-                    measure_index += 1
                 elif line.startswith('#NOTES:'):
-                    measure_index = 0                   
+                    flag = True
+            if flag:
```
Comment (cpuguy96) [inline smfile_parser.py]: Change to else

File: smfile_parser.py [hunk 3]
```diff
@@ -1,103 +1,95 @@
...
+                if not (line.startswith(' ') or line.startswith('/') or line.startswith('\n') or line.startswith('#')):
+                    data['notes'].append(sub('4', '1', sub('[MKLF]', '0', line)).rstrip('\n'))
+                if line.startswith('#NOTES:'):
                     next(f)
                     next(f)
-                    step_dict['types'].append("DIFFICULTY ")
-                    step_dict['notes'].append(next(f).lstrip(' ').rstrip(':\n'))
-    return step_dict
+                    data['notes'].append('DIFFICULTY ' + next(f).lstrip(' ').rstrip(':\n'))
```
Comment (cpuguy96) [inline smfile_parser.py]: This breaks the format of parsing 1 line per loop. Can this be changed to keep that logic?

File: smfile_parser.py [hunk 4]
```diff
@@ -1,103 +1,95 @@
...
-                    step_dict['notes'].extend(line_timing)
-                    measure.clear()
-                    if line.startswith(';'):
-                        output_file(new_file, step_dict, output_dir)
-                    measure_index += 1
                 elif line.startswith('#NOTES:'):
-                    measure_index = 0                   
+                    flag = True
+            if flag:
+                if not (line.startswith(' ') or line.startswith('/') or line.startswith('\n') or line.startswith('#')):
```
Comment (cpuguy96) [inline smfile_parser.py]: This can be combined into 1 statement:
`line.startswith(' ', '/', '\n', '#')`

File: test_parser.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        measure = [True, False, False, True]
+        measure_index = 2
+        bpm = 200.000
+        offset = 0.009
+
+        timings = parse.calculate_timing(measure, measure_index, bpm, offset)
+        self.assertTrue(all(float(time) >= 0 for time in timings))
+        self.assertEqual(timings, ["2.391", "3.291"])
+
+    #def test_parse_sm(self):
```
Comment (cpuguy96) [inline test_parser.py]: Add unit tests for each function and add failure cases

File: smfile_parser.py [hunk 6]
```diff
@@ -1,103 +1,95 @@
...
-                    step_dict['notes'].append(next(f).lstrip(' ').rstrip(':\n'))
-    return step_dict
+                    data['notes'].append('DIFFICULTY ' + next(f).lstrip(' ').rstrip(':\n'))
+
+    return data
+
+def clean_data(data):
+    data['title']  = data['title'].lstrip('#TILE').lstrip(':').rstrip(';\n')
+    data['bpm']    = float(data['bpm'].lstrip('#BPMS:.0').lstrip('=').rstrip(';\n'))
+    data['offset'] = float(data['offset'].lstrip('#OFSET:').rstrip(';\n'))
```
Comment (cpuguy96) [inline smfile_parser.py]: Why move all this logic outside the parse function? This function is only being called once.

File: smfile_parser.py [hunk 7]
```diff
@@ -1,103 +1,95 @@
...
+    return timing
+
+def generate_timing(data):
+    measure = []
+    notes = []
+    position = 0
+
+    for line in data['notes']:
+        if line[0].isdigit():
+            check = True if any((c in set('123456789')) for c in line) else False
```
Comment (cpuguy96) [inline smfile_parser.py]: Rename

File: smfile_parser.py [hunk 8]
```diff
@@ -48,56 +47,51 @@ def parse_sm(sm_file, new_file, output_dir):
...
-                if line[0].isdigit():
-                    check = True if any((c in set('123456789')) for c in line) else False
-                    if check:
-                        measure.append(True)
-                        step_dict['types'].append(convert_note(line.rstrip('\n')))
-                    else:
-                        measure.append(False)
-                elif line.startswith(',') or line.startswith(';'):
-                    line_timing = calculate_timing(measure, measure_index, step_dict['BPM'], step_dict['offset'])
+                    step_dict['notes'].append('DIFFICULTY ' + next(f).lstrip(' ').rstrip(':\n'))
```
Comment (cpuguy96) [inline smfile_parser.py]: It's possible this can cause EOF exceptions if the file is not properly formatted.

Comment (cpuguy96) [reply smfile_parser.py]: Also, why add the difficulty to notes? You can make 'notes' be a list of dictionaries with the keys being the difficulty name and the value being the measures.

File: smfile_parser.py [hunk 9]
```diff
@@ -48,56 +47,51 @@ def parse_sm(sm_file, new_file, output_dir):
...
                 elif line.startswith('#OFFSET:'):
-                    step_dict['offset'] = float(line.lstrip('#OFFSET').lstrip(':').rstrip(';\n'))
-                elif line.startswith('#STOPS:') and line.rstrip('\n') != "#STOPS:;":
+                    step_dict['offset'] = float(split(':', line)[-1].rstrip(';'))
+                elif line.startswith('#STOPS:') and line != "#STOPS:;":
                     raise ValueError('Stop detected')
-
                 elif line.startswith('#NOTES:'):
-                    flag     = True
+                    read_notes     = True
```
Comment (cpuguy96) [inline smfile_parser.py]: I realized now you can just split by the first `:` to get the type of metadata. Afterwards, you can parse the value based on what type of metadata it is.

**Review by cpuguy96 — COMMENTED** [review]
Add comments

File: smfile_parser.py [hunk 10]
```diff
@@ -34,70 +35,67 @@ def convert_note(line):
...
-                    if check:
-                        measure.append(True)
-                        step_dict['types'].append(convert_note(line.rstrip('\n')))
-                    else:
-                        measure.append(False)
-                elif line.startswith(',') or line.startswith(';'):
-                    line_timing = calculate_timing(measure, measure_index, step_dict['BPM'], step_dict['offset'])
-                    step_dict['notes'].extend(line_timing)
+                    current_difficulty = next(f).lstrip(' ').rstrip(':\n')
+                elif line.startswith((',', ';')):
```
Comment (cpuguy96) [inline smfile_parser.py]: Add check if `current_difficulty` is empty

File: smfile_parser.py [hunk 11]
```diff
@@ -34,70 +35,67 @@ def convert_note(line):
...
 
 def calculate_timing(measure, measure_index, bpm, offset):  #calculate time in seconds for each line
-    measure_seconds = 4 * 60/bpm    #length of measure in seconds   
+    measure_seconds = 4 * 60/bpm    #length of measure in seconds
     note_256        = measure_seconds/256   #length of each 1/256th note in the measure in seconds
     measure_timing  = measure_seconds * measure_index   #accumulated time from previous measures
     fraction_256    = 256/len(measure)  #number of 1/256th notes per beat: 1/2nd = 128, 1/4th = 64, etc
 
-    line_timing = [str(i * note_256 * fraction_256 + measure_timing - offset) for i, is_set in enumerate(measure) if is_set]
+    line_timing = [measure[i] + ' ' + str(i * note_256 * fraction_256 + measure_timing - offset) for i, is_set in enumerate(measure) if is_set != None]
```
Comment (cpuguy96) [inline smfile_parser.py]: Rename to make it clear it contains measure and timing

File: smfile_parser.py [hunk 12]
```diff
@@ -34,70 +35,70 @@ def convert_note(line):
...
-                        output_file(new_file, step_dict, output_dir)
                     measure_index += 1
-                elif line.startswith('#NOTES:'):
-                    measure_index = 0                   
-                    next(f)
-                    next(f)
-                    step_dict['types'].append("DIFFICULTY ")
-                    step_dict['notes'].append(next(f).lstrip(' ').rstrip(':\n'))
+                elif line and not line.startswith(' '): # individual notes
+                    line = convert_note(line)
```
Comment (cpuguy96) [inline smfile_parser.py]: Rename

**Comment by jhaco** [comment]
I've found an edge case where an .sm file had multiple BPMs placed on their own line, rather than all on one line. This led to a few extra files being "successfully parsed" when they shouldn't be. The distinguishing factor is ' ; ' at the end, so I'm going to rewrite the parsing to check for that before the script decides it has read all the information for BPM (and other info).

**Comment by cpuguy96** [comment]
Good catch! Do you know if all the metadata is structured as `#type: value;`? If so, you can read starting at a # and then parse until you find a ;

**Comment by cpuguy96** [comment]
Looks good. Just add unit tests.

**Comment by jhaco** [comment]
Merging and moving activity to SMDataTools

**Pull request merged**